[sharded_tensor] refactor sharded_tensor file structure

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66351
* #64481
* __->__ #67199

This PR refactors _sharded_tensor package so that it splits from api.py, and add different components to make it more modularized, this will also help us resolve circular dependency due to increasing code size and better organize the package:

* api.py: sharded tensor APIs
* metadata.py: Metadata definition for ShardedTensors
* shard.py: Shard definition for ShardedTensor
* utils.py: utility functions for validation, etc.

Differential Revision: [D31904249](https://our.internmc.facebook.com/intern/diff/D31904249/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

Commit:
[sharded_tensor] refactor sharded_tensor file structure

This PR refactors _sharded_tensor package so that it splits from api.py, and add different components to make it more modularized, this will also help us resolve circular dependency due to increasing code size and better organize the package:

* api.py: sharded tensor APIs
* metadata.py: Metadata definition for ShardedTensors
* shard.py: Shard definition for ShardedTensor
* utils.py: utility functions for validation, etc.

Differential Revision: [D31904249](https://our.internmc.facebook.com/intern/diff/D31904249/)

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] refactor sharded_tensor file structure"


This PR refactors _sharded_tensor package so that it splits from api.py, and add different components to make it more modularized, this will also help us resolve circular dependency due to increasing code size and better organize the package:

* api.py: sharded tensor APIs
* metadata.py: Metadata definition for ShardedTensors
* shard.py: Shard definition for ShardedTensor
* utils.py: utility functions for validation, etc.

Differential Revision: [D31904249](https://our.internmc.facebook.com/intern/diff/D31904249/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] refactor sharded_tensor file structure"


This PR refactors _sharded_tensor package so that it splits from api.py, and add different components to make it more modularized, this will also help us resolve circular dependency due to increasing code size and better organize the package:

* api.py: sharded tensor APIs
* metadata.py: Metadata definition for ShardedTensors
* shard.py: Shard definition for ShardedTensor
* utils.py: utility functions for validation, etc.

Differential Revision: [D31904249](https://our.internmc.facebook.com/intern/diff/D31904249/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

pritamdamania87 reviewed (approved):
Nice clean refactor!

Commit:
Update on "[sharded_tensor] refactor sharded_tensor file structure"


This PR refactors _sharded_tensor package so that it splits from api.py, and add different components to make it more modularized, this will also help us resolve circular dependency due to increasing code size and better organize the package:

* api.py: sharded tensor APIs
* metadata.py: Metadata definition for ShardedTensors
* shard.py: Shard definition for ShardedTensor
* utils.py: utility functions for validation, etc.

Differential Revision: [D31904249](https://our.internmc.facebook.com/intern/diff/D31904249/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]
